### ISSUE TITLE
Current status of the transition to `NSURLSession` or AFNetworking 2.0

### ISSUE THREAD

**Issue by AlexanderKaraberov** [RestKit/RestKit#2371]
I'm sorry if this issue was already discussed/is discussing, then feel free to close mine and point me to the appropriate discussion thread. 
<p> I'm interested: what are the future plans for networking module of the framework? Two years ago Blake said he want to get rid of `AFNetworking` dependency at all and just use `NSURLSession` for all the networking logic in the framework. I think this is a very reasonable idea. But as I see now, old AFNetworking 1.x code is still used. Can someone tell me about the future plans and current status of the transition process. Because using an outdated version of `AFNetworking` can be a barrier which can prevent using `RestKit` in modern iOS project.

**Comment by segiddins** [comment]
Still waiting on a pull request!

**Comment by wainglaister** [comment]
Take a look at https://github.com/oligriffiths/RestKit/tree/feature/test_conformance?files=1 which represents the current working version.
Needs more testing and then we'll need to merge changes around. It isn't quite ready to go yet.

**Comment by tinora** [comment]
Hi @segiddins, @wainglaister, et al!
Let me pls follow up on this question. Oli Griffiths has made some really good progress regarding the AFNetworking-free version of RestKit. So far his work is usable and works without major problems. It doesn't completely break up the dependency between RestKit and AFNetworking though. RestKit still depends on the 'AFNetworking/Serialization'.  Unfortunately Oli's work is still not on the master and I haven't seen any progress from him for quite some month now. Last year a colleague of mine and I promised to invest some time in order to help getting the AFNetworking-free transition completed. We haven't really delivered (besides some bug fixes on Oli's branch) on our promises yet. But as we are still highly interested in getting the AFNetworking-free version into the master and as Oli seems to has stopped his effort we will now (and have already started) to invest more time into the project. Our time is limited but we will try to invest at least a day per week till this is done. Now after even more promises ;) our plan on how to continue:

1. We have merged the branch we are working on with the latest RestKit master in order to be up-to-date
2. We got the tests to run, found and fixed a couple of failed tests (some more remain to be fixed)
3. We would copy the 'AFNetworking/Serialization' (including the corresponding unit tests) files into RestKit, prepending them with a 'RK' prefix in order to avoid name clashes when using AFNetworking and RestKit together. I think there is no point in recreating the AFNetworking serializer functionality from scratch. Besides that parts of AFNetworking have been copied into RestKit (see ../Code/Support/RKURLEncodedSerialization.m) already. 
4. Finish or rework some of the code Oli has written. The design and implementation of RKHTTPClient still has some rough and unfinished edges. Furthermore it need to be discussed if and how to replace the reachability functionality AFHTTPClient had. 

No plan survives first contact with the enemy, so I wouldn't plan to much in advance in order to be flexible. The above mentioned points are the ones we are currently working on. 
I'd be happy to get feedback or help. I think we all want to bring this effort to an end eventually. 

Regards,
Tino

**Comment by valeriomazzeo** [comment]
Thanks @tinora, that's a very good news.

I have been sporadically look into this during the past week and the main issue was the inability to get anything to compile when refactoring the Network subspec.

If you have a look at this branch / commit 
https://github.com/RestKit/RestKit/commit/75a52e99329fb825ab5c0568e4fca64d6c4c6bfb

I have literally managed to cleanup the whole Network layer, that was a bit entangled here and there with other subspaces (especially in the tests).
This provides a base workbench to start adding incrementally all the components needs for the network layer without worrying about project not compiling and tests not running that (at least for me) it's very frustrating.

Current status is:

* Project deployment target is up to 8.0 (so we can use URITemplate)
* Project compile with Cocoapods 1.0.1
* Network subspec has been completely removed
* Tests are passing
----
* Example projects are obviously not compiling
* There's a problem with `pod lib lint` complaining about some missing files

In terms of what direction to take, I have been looking at removing completely AFNetworking, but it seems to me that we are going to end up reinventing the wheel, that's why, the way I was going to approach this was to create an abstraction layer between RestKit and whatever client we may want to use now or in the future that should be easily pluggable through the implementation of simple adapters.

It would be interesting to see what are your thoughts, and hopefully my branch could help you with a clean start.

**Comment by tinora** [comment]
Thanks for your response @valeriomazzeo.  I've got a couple of questions and comments. 
1. Did I get you right that you have removed the whole network layer including tests from your branch? If this is the case and though I understand and welcome what you are aiming at  I'm skeptical that this is the most pragmatic way right now if we finally want to make a step forward in getting RestKit AFNetworking-free in a reasonable time frame (weeks and not months). It has been tried for a couple of years now to get rid of AFNetworking without success. Without doubt your approach to abstract out the network layer is valid and right. I think however that it will cost a lot more time to agree on a suitable API etc. Or how long would you think it will take you to complete this effort?
My first goal would be to break the last connection between RestKit and AFNetworking as soon as possible and get this state into the master. So everybody can start using the official RestKit version again together with whatever AFNetworking version. I roughly mentioned in my previous post how I would approach this. I don't claim that this is technically the wisest approach and it certainly generates some rework if we'd finally aim at refactoring the whole network layer anyway. It is more driven by pragmatic aspects as said.
Once we are there trying to refactor the network layer is certainly a laudable goal. One that can be targeted with little less time pressure. Would you agree?

2. Bumping up the iOS/OSX deployment target is a good move in my opinion. I think iOS 8 is a reasonable target. I don't think anybody will start new iOS project below iOS 8 today.

What do you think?

Regards
Tino

**Comment by tinora** [comment]
@mkoehnke (my colleague) has put a lot of effort into this project already so I think it makes sense to include him into this discussion.

**Comment by valeriomazzeo** [comment]
Correct. Looking at what has been done in the other branches `porting-to-afnetworking2` and `objectmapper2.0` it looks to me that the whole network layer was meant to have interface breaking changes anyway.

> Without doubt your approach to abstract out the network layer is valid and right. I think however that it will cost a lot more time to agree on a suitable API etc. Or how long would you think it will take you to complete this effort?

The idea here is to maintain, as much as it's humanly possible and it make sense, the same interface for classes like `RKRouter`, `RKRouteSet`, `RKRoute`, `RKResponseDescriptor`, `RKRequestDescriptor`. also trying re-using the tests as well. This would probably take few days of development, where the main target is in my opinion replace `SOCKit` with `URITemplate`.

The problem with the code in the repository you mentioned, is that it's still using `NSURLConnection` whether I'd rather migrate to `NSURLSession` with or without `AFNetworking`. In the `NSURLConnection` world seems that `AFNetworking` only adds some real value on the serialisers aspect though regarding the `RestKit` project.

I don't think the object manager implementation, as it stands right now, using `NSOperation` is suitable anymore, due to the use of `dataTask` which behaves completely differently.

Any idea / plan is welcome, please contact me by email though, as a quick chat maybe solve more doubts quickly.

**Comment by tinora** [comment]
Our current work is based on Oli Griffiths branch, see https://github.com/apploft/RestKit 
Doing interface breaking changes is not a goal on this branch. 

> ...the main target is in my opinion replace SOCKit with URITemplate.

I won't disagree that this might be a useful goal in itself. But do you think it is a precondition on the way toward an AFNetworking-free version of RestKit or is it rather a separate goal?

> "...The problem with the code in the repository you mentioned, is that it's still using NSURLConnection..." 

On our branch NSURLSession will already be used. The RestKit and RestKitFramework targets are compiling on our branch and most of the tests do also run (some do fail and need to be fixed). 

> "Any idea / plan is welcome..."

As said my main goal right now would be to finish the AFNetworking-free RestKit version and get it integrated into the master. I would ignore everything that is not absolutely necessary on the way towards this goal. I would further accept a technical solution which is not perfect but takes us a real step forward. This whole issue is lingering for too long already.

**Comment by valeriomazzeo** [comment]
That was due to carry on the work that was started with `porting-to-afnetworking2` and `objectmapper2.0`. If you guys are almost there and it's ready to go then I agree with you on completing it  before hand without breaking change to the interface.

To be honest I had a look at Oli Griffiths branch and not at the `apploft` one so I was probably missing something.

As long as, you are able to merge in the `develop` branch and get's everything working and documented then I don't see any harm in going this direction.

Happy to help, with anything :)

**Comment by tinora** [comment]
Thanks @valeriomazzeo 
Feel free to check out our branch and let us know if you have problems, suggests etc.
As said, there are still a couple of tests that fail at the moment. Any help getting them to succeed is appreciated. I can invest into the project first next week again.

Regards
Tino

**Comment by raxelz** [comment]
Guys I am using this fork instead of main RestKit dev branch, and it works like a charm but did any of you get to test if it passes IPV6?
I'm relatively a fresh developer, I'm trying out Apple's IPV6 compliance test and my requests are failing under IPV6 simulated network... even though this does not use any NSURLConnections or hardcore IPV4 etc...
If I connect to my normal network everything works fine o_o
Any chance you can test it or help me out? I super appreciate you guys moving this to NSURLSession, I know its a lot of work!
